rule/import-shadowing.go

```diff
@@ -0,0 +1,96 @@
+package rule
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"strings"
+
+	"github.com/mgechev/revive/lint"
+)
+
+// ImportShadowingRule lints given else constructs.
+type ImportShadowingRule struct{}
+
+// Apply applies the rule to given file.
+func (r *ImportShadowingRule) Apply(file *lint.File, _ lint.Arguments) []lint.Failure {
+	var failures []lint.Failure
+
+	importNames := map[string]struct{}{}
+	for _, imp := range file.AST.Imports {
+		importNames[getName(imp)] = struct{}{}
+	}
+
+	fileAst := file.AST
+	walker := importShadowing{
+		importNames: importNames,
+		onFailure: func(failure lint.Failure) {
+			failures = append(failures, failure)
+		},
+		alreadySeen: map[*ast.Object]struct{}{},
+	}
+
+	ast.Walk(walker, fileAst)
+
+	return failures
+}
+
+// Name returns the rule name.
+func (r *ImportShadowingRule) Name() string {
+	return "import-shadowing"
+}
+
+func getName(imp *ast.ImportSpec) string {
+	const pathSep = "/"
+	const strDelim = `"`
+	if imp.Name != nil {
+		return imp.Name.Name
+	}
+
+	path := imp.Path.Value
+	i := strings.LastIndex(path, pathSep)
+	if i == -1 {
+		return strings.Trim(path, strDelim)
+	}
+
+	return strings.Trim(path[i+1:], strDelim)
+}
+
+type importShadowing struct {
+	importNames map[string]struct{}
+	onFailure   func(lint.Failure)
+	alreadySeen map[*ast.Object]struct{}
+}
+
+func (w importShadowing) Visit(n ast.Node) ast.Visitor {
+	switch n := n.(type) {
+	case *ast.AssignStmt:
+		if n.Tok == token.DEFINE {
+			return w // analyze variable declarations of the form id := expr
+		}
+
+		return nil
+	case *ast.CallExpr, *ast.ImportSpec, *ast.KeyValueExpr, *ast.ReturnStmt, *ast.SelectorExpr, *ast.StructType:
```

Comment:
This is for optimization to not traverse the subtree?

Reply:
Is to avoid failures on each reference to a shadowing identifier.

Reply:
Isn't this handled by the `alreadySeen` property?

Reply:
My previuos message was not clear. While `alreadySeen` avoids failing multiple times on a shadowing identifier, the case
```suggestion
	case *ast.CallExpr, *ast.ImportSpec, *ast.KeyValueExpr, *ast.ReturnStmt, *ast.SelectorExpr, *ast.StructType:
```
Avoids to analyze subtrees that can raise _false positives_; for example:

* skips analysis of the `import` sub-tree: we already have the list of imports
* skips analysis of key-val expressions (`{key:value}`): ids of such expressions, even the same of an import name, do not shadow the import name
* skips analysis of selector expressions (`anId.otherId`): because if `anId`shadows an import name, it was already detected, and `otherId` does not shadows the import name
* skips analysis of struct type because struct fields can not shadow an import name

skipping analysis of returns is an optimization but for the sake of clarity we can remove it from the case expression

Reply:
Thanks for the clarification! Would you add this as a comment in the rule? It'll be very useful to make sure we don't lose this information.

Reply:
done